presto-main/src/test/java/com/facebook/presto/operator/TestGroupByHash.java

```diff
@@ -233,4 +242,46 @@ public void testForceRehash()
             assertTrue(groupByHash.contains(i, new Page(valuesBlock, hashBlock), CONTAINS_CHANNELS));
         }
     }
+
+    @Test(dataProvider = "dataType")
+    public void testReserveMemory(Type type)
+            throws Exception
+    {
+        ImmutableList.Builder<Long> memoryAllocations = ImmutableList.builder();
+
+        // Create a page with positionCount >> expected size of groupByHash
+        int length = 1_000_000;
+        Block valuesBlock;
+        if (type == VARCHAR) {
+            valuesBlock = createStringSequenceBlock(0, length);
+        }
+        else if (type == BIGINT) {
+            valuesBlock = createLongSequenceBlock(0, length);
+        }
+        else {
+            throw new IllegalArgumentException("unsupported data type");
+        }
+        Block hashBlock = getHashBlock(ImmutableList.of(type), valuesBlock);
+
+        // Create group by hash with extremely small size
+        GroupByHash groupByHash = createGroupByHash(
+                TEST_SESSION,
+                ImmutableList.of(type),
+                new int[] {0},
+                Optional.of(1),
+                4,
+                JOIN_COMPILER,
+                x -> {
+                    memoryAllocations.add(x);
+                    return true;
+                });
+        groupByHash.addPage(new Page(valuesBlock, hashBlock));
+
+        // Ensure we double the size of reserved memory
+        List<Long> memorySizes = memoryAllocations.build();
+        assertTrue(memorySizes.size() > 1);
+        for (int i = 0; i < memorySizes.size() - 1; i++) {
+            assertEquals(memorySizes.get(i) * 2, (long) memorySizes.get(i + 1));
+        }
```

Comment:
Replace this for-loop with a simple assertEquals:
`
assertEquals(memorySizes, ImmutableList.of(.......));
`

If necessary, reduce `length`.